### PR TITLE
fix a bug in k2.ragged.normalize_scores.

### PR DESCRIPTION
**Pull request by csukuangfj**
If k2.ragged.RaggedFloat is constructed from a shape and a value,
its scores should be set to the given value to make autograd work.

### CODE REVIEW

**Comment by csukuangfj** [comment]
There already exists `k2.Fsa.detach()`
https://github.com/k2-fsa/k2/blob/75f9378c3bae8c01a280a32cb4c2f6afb938951a/k2/python/k2/fsa.py#L887

so I am removing the one introduced a few days ago.

**Comment by danpovey** [comment]
looks OK to me..

File: k2/python/k2/ragged/tensor.py [hunk 1]
```diff
@@ -42,14 +42,18 @@ def __init__(self,
...
         elif isinstance(ragged, _k2.RaggedShape):
             assert values is not None
             ragged = _k2.RaggedFloat(ragged, values)
 
         assert isinstance(ragged, _k2.RaggedFloat)
 
         self.ragged = ragged
-        self._scores = ragged.values()
+        if values is not None:
+            self._scores = values
```
Comment (danpovey) [inline k2/python/k2/ragged/tensor.py:54]: Wait, why does RaggedFloat have a _scores member?

Comment (danpovey) [reply k2/python/k2/ragged/tensor.py:54]: ... it is supposed to be a generic interface, not tied into class Fsa.

Comment (csukuangfj) [reply k2/python/k2/ragged/tensor.py:54]: > Wait, why does RaggedFloat have a _scores member?

While doing `normalize_scores`, if we use `_k2.RaggedFloat` directly it is not possible to register the
operation into the autograd graph since there is no `torch.Tensor`. 

`k2.ragged.RaggedFloat` is a wrap around `_k2.RaggedFloat` and the field `_scores` is to make
autograd work.

https://github.com/k2-fsa/k2/blob/75f9378c3bae8c01a280a32cb4c2f6afb938951a/k2/python/k2/ragged/autograd.py#L15-L16

For the autograd to work, we need to pass a `torch.Tensor` to it. One alternative is to let the user pass a tensor
while calling `normalize_scores`.

Comment (csukuangfj) [reply k2/python/k2/ragged/tensor.py:54]: > ... it is supposed to be a generic interface, not tied into class Fsa.

The class `RaggedFloat` is not tied to `class Fsa`; it is a wrapper around `_k2.RaggedFloat`.

Comment (danpovey) [reply k2/python/k2/ragged/tensor.py:54]: I think the name _values would be better than _scores; we don't want to imply the values have any specific meaning.

Comment (csukuangfj) [reply k2/python/k2/ragged/tensor.py:54]: will change it.

Comment (danpovey) [reply k2/python/k2/ragged/tensor.py:54]: And please make sure the documentation of class RaggedFloat is as self-contained as possible and explains anything that needs to be explained about autograd...  (it may already be; just a reminder).

Comment (csukuangfj) [reply k2/python/k2/ragged/tensor.py:54]: I am adding more documentation for it.

**Comment by csukuangfj** [comment]
Ready for review.

File: k2/python/k2/fsa.py [hunk 2]
```diff
@@ -1029,44 +1029,23 @@ def from_openfst(cls, s: str, acceptor: bool = True) -> 'Fsa':
         arcs, aux_labels = _k2.fsa_from_str(s, acceptor, True)
         return Fsa(arcs, aux_labels=aux_labels)
 
-    def set_scores_stochastic_(self) -> None:
-        '''Set `scores` to random numbers.
+    def set_scores_stochastic_(self, scores) -> None:
+        '''Normalize the given `scores` and assign it to self.
```
Comment (danpovey) [inline k2/python/k2/fsa.py:1033]: I'd write this as:
   Normalize the given `scores` and assign the result to self.scores.
 Args:
     scores:   Tensor of scores of dtype torch.float32, and shape equal to self.scores.shape
            (one axis).  Will be normalized so the sum, after exponentiating, of the scores 
            leaving each state that has at least one arc leaving it is 1.

File: k2/python/k2/fsa.py [hunk 3]
```diff
@@ -1029,44 +1029,23 @@ def from_openfst(cls, s: str, acceptor: bool = True) -> 'Fsa':
...
-
-        for name, value in self.named_non_tensor_attr():
-            setattr(ans, name, value)
+        ragged_scores = k2.ragged.RaggedFloat(
+            self.arcs.shape().to(scores.device), scores)
+        ragged_scores = k2.ragged.normalize_scores(ragged_scores)
 
-        ans.scores = self.scores.detach()
-        return ans
+        # Note we use `to` here since `scores` and `self.scores` may not
```
Comment (danpovey) [inline k2/python/k2/fsa.py:1049]: I find it odd that scores and self.scores might not be on the same device, surely we should ensure that they always are?

Comment (csukuangfj) [reply k2/python/k2/fsa.py:1049]: It happens in our current ASG training:

- model.P_scores is on CUDA
- P is on CPU since `k2.intersect` works only on CPU

So in this case `model.P_scores` and `P.scores` are on different devices.

File: k2/python/k2/fsa.py (hunk 2 above)
Comment (csukuangfj) [reply k2/python/k2/fsa.py]: thanks, changed.

File: k2/python/k2/fsa.py (hunk 3 above)
Comment (csukuangfj) [reply k2/python/k2/fsa.py:1053]: >  surely we should ensure that they always are?

I assume that log-sum per sub list runs faster on CUDA; otherwise, we can require that `scores.device == self.device`.

**Comment by csukuangfj** [comment]
Merging. If there are any objections, please leave your comments here and I will resolve them.
